### PR TITLE
Pip: Get package metadata from python-inspector

### PR DESCRIPTION
**Pull request by mnonnenmacher**
Get the metadata of Python packages from the python-inspector output. There are two follow-ups for this PR planned:

* Change the default values for `operatingSystem` and `pythonVersion` to use the current operating system and the installed Python version.
* Update the Dockerfile with the new reduced requirements for the Python installation.

### CODE REVIEW

File: analyzer/src/funTest/assets/projects/external/spdx-tools-python-expected-output.yml [hunk 1]
```diff
@@ -135,27 +134,23 @@ packages:
...
     spdx_expression: "BSD-3-Clause"
-    mapped:
-      BSD License: "BSD-3-Clause"
-  description: "RDFLib is a Python library for working with RDF, a simple yet powerful\
-    \ language for representing information."
+  description: "RDFLib is a Python library for working with"
   homepage_url: "https://github.com/RDFLib/rdflib"
   binary_artifact:
-    url: "https://files.pythonhosted.org/packages/d0/6b/6454aa1db753c0f8bc265a5bd5c10b5721a4bb24160fb4faf758cf6be8a1/rdflib-5.0.0-py3-none-any.whl"
+    url: ""
```
Comment (sschuberth) [inline analyzer/src/funTest/assets/projects/external/spdx-tools-python-expected-output.yml:145]: Odd, why's this missing?

File: analyzer/src/funTest/assets/projects/synthetic/pip-expected-output.yml [hunk 2]
```diff
@@ -34,24 +34,24 @@ packages:
 - id: "PyPI::click:6.7"
   purl: "pkg:pypi/click@6.7"
   authors:
-  - "Armin Ronacher"
+  - "Armin Ronacher <armin.ronacher@active-4.com>"
   declared_licenses:
-  - "BSD License"
+  - ""
```
Comment (sschuberth) [inline analyzer/src/funTest/assets/projects/synthetic/pip-expected-output.yml:39]: We should never store an empty license. This needs to be fixed in the implementation.

File: analyzer/src/funTest/assets/projects/synthetic/pip-expected-output.yml [hunk 3]
```diff
@@ -99,25 +97,24 @@ packages:
 - id: "PyPI::itsdangerous:0.24"
   purl: "pkg:pypi/itsdangerous@0.24"
   authors:
-  - "Armin Ronacher"
+  - "Armin Ronacher <armin.ronacher@active-4.com>"
   declared_licenses:
-  - "BSD License"
+  - "UNKNOWN"
```
Comment (sschuberth) [inline analyzer/src/funTest/assets/projects/synthetic/pip-expected-output.yml:102]: This looks also wrong.

File: analyzer/src/funTest/assets/projects/synthetic/pip-expected-output.yml [hunk 4]
```diff
@@ -99,25 +97,24 @@ packages:
...
     unmapped:
-    - "BSD License"
-  description: "Various helpers to pass trusted data to untrusted environments and\
-    \ back."
+    - "UNKNOWN"
+  description: "It's Dangerous"
   homepage_url: "http://github.com/mitsuhiko/itsdangerous"
   binary_artifact:
-    url: "https://files.pythonhosted.org/packages/dc/b4/a60bcdba945c00f6d608d8975131ab3f25b22f2bcfe1dab221165194b2d4/itsdangerous-0.24.tar.gz"
+    url: ""
```
Comment (sschuberth) [inline analyzer/src/funTest/assets/projects/synthetic/pip-expected-output.yml:109]: Why's this missing?

File: analyzer/src/funTest/assets/projects/synthetic/pip-expected-output.yml [hunk 5]
```diff
@@ -164,27 +158,25 @@ packages:
...
     spdx_expression: "BSD-3-Clause"
     mapped:
       BSD: "BSD-3-Clause"
-      BSD License: "BSD-3-Clause"
-  description: "Implements a XML/HTML/XHTML Markup safe string for Python"
+  description: "MarkupSafe"
   homepage_url: "http://github.com/pallets/markupsafe"
   binary_artifact:
-    url: "https://files.pythonhosted.org/packages/4d/de/32d741db316d8fdb7680822dd37001ef7a448255de9699ab4bfcbdf4172b/MarkupSafe-1.0.tar.gz"
+    url: ""
```
Comment (sschuberth) [inline analyzer/src/funTest/assets/projects/synthetic/pip-expected-output.yml:171]: Why's this missing?

File: analyzer/src/funTest/assets/projects/synthetic/pipenv-expected-output.yml [hunk 6]
```diff
@@ -30,24 +30,24 @@ packages:
 - id: "PyPI::click:6.7"
   purl: "pkg:pypi/click@6.7"
   authors:
-  - "Armin Ronacher"
+  - "Armin Ronacher <armin.ronacher@active-4.com>"
   declared_licenses:
-  - "BSD License"
+  - ""
```
Comment (sschuberth) [inline analyzer/src/funTest/assets/projects/synthetic/pipenv-expected-output.yml:35]: Again, empty licenses should be omitted from here.

File: analyzer/src/funTest/assets/projects/synthetic/pipenv-expected-output.yml [hunk 7]
```diff
@@ -95,25 +93,24 @@ packages:
 - id: "PyPI::itsdangerous:0.24"
   purl: "pkg:pypi/itsdangerous@0.24"
   authors:
-  - "Armin Ronacher"
+  - "Armin Ronacher <armin.ronacher@active-4.com>"
   declared_licenses:
-  - "BSD License"
+  - "UNKNOWN"
```
Comment (sschuberth) [inline analyzer/src/funTest/assets/projects/synthetic/pipenv-expected-output.yml:98]: This looks wrong.

File: analyzer/src/funTest/assets/projects/synthetic/pipenv-expected-output.yml [hunk 8]
```diff
@@ -95,25 +93,24 @@ packages:
...
     unmapped:
-    - "BSD License"
-  description: "Various helpers to pass trusted data to untrusted environments and\
-    \ back."
+    - "UNKNOWN"
+  description: "It's Dangerous"
   homepage_url: "http://github.com/mitsuhiko/itsdangerous"
   binary_artifact:
-    url: "https://files.pythonhosted.org/packages/dc/b4/a60bcdba945c00f6d608d8975131ab3f25b22f2bcfe1dab221165194b2d4/itsdangerous-0.24.tar.gz"
+    url: ""
```
Comment (sschuberth) [inline analyzer/src/funTest/assets/projects/synthetic/pipenv-expected-output.yml:105]: Why's this missing?

File: analyzer/src/funTest/assets/projects/synthetic/pipenv-expected-output.yml [hunk 9]
```diff
@@ -162,27 +156,25 @@ packages:
...
     spdx_expression: "BSD-3-Clause"
     mapped:
       BSD: "BSD-3-Clause"
-      BSD License: "BSD-3-Clause"
-  description: "Implements a XML/HTML/XHTML Markup safe string for Python"
+  description: "MarkupSafe"
   homepage_url: "http://github.com/pallets/markupsafe"
   binary_artifact:
-    url: "https://files.pythonhosted.org/packages/4d/de/32d741db316d8fdb7680822dd37001ef7a448255de9699ab4bfcbdf4172b/MarkupSafe-1.0.tar.gz"
+    url: ""
```
Comment (sschuberth) [inline analyzer/src/funTest/assets/projects/synthetic/pipenv-expected-output.yml:169]: Why's this missing?

File: analyzer/src/funTest/assets/projects/synthetic/poetry-expected-output.yml [hunk 10]
```diff
@@ -26,26 +26,23 @@ packages:
...
   binary_artifact:
     url: "https://files.pythonhosted.org/packages/9d/fb/886e8ec7862989afc0c35d15813b6c665fe134cc6027cdde2fa300abe9a2/graphviz-0.19.1-py3-none-any.whl"
     hash:
-      value: "c0e5b8e7497c70c0c4cd83cc8c3bdedf"
-      algorithm: "MD5"
+      value: "f34088c08be2ec16279dfa9c3b4ff3d1453c5c67597a33e2819b000e18d4c546"
+      algorithm: "SHA-256"
   source_artifact:
-    url: "https://files.pythonhosted.org/packages/f6/14/aa3ec10e4ab1524ba69f4742ef9cfa01fd668d0840afe5d5e6f708a95031/graphviz-0.19.1.zip"
+    url: ""
```
Comment (sschuberth) [inline analyzer/src/funTest/assets/projects/synthetic/poetry-expected-output.yml:42]: Why's this missing?

File: analyzer/src/funTest/assets/projects/synthetic/python-inspector-expected-output.yml [hunk 11]
```diff
@@ -30,24 +30,24 @@ packages:
 - id: "PyPI::click:6.7"
   purl: "pkg:pypi/click@6.7"
   authors:
-  - "Armin Ronacher"
+  - "Armin Ronacher <armin.ronacher@active-4.com>"
   declared_licenses:
-  - "BSD License"
+  - ""
```
Comment (sschuberth) [inline analyzer/src/funTest/assets/projects/synthetic/python-inspector-expected-output.yml:35]: Again, empty licenses should be omitted from here.

File: analyzer/src/funTest/assets/projects/synthetic/python-inspector-expected-output.yml [hunk 12]
```diff
@@ -95,25 +93,24 @@ packages:
...
     unmapped:
-    - "BSD License"
-  description: "Various helpers to pass trusted data to untrusted environments and\
-    \ back."
+    - "UNKNOWN"
+  description: "It's Dangerous"
   homepage_url: "http://github.com/mitsuhiko/itsdangerous"
   binary_artifact:
-    url: "https://files.pythonhosted.org/packages/dc/b4/a60bcdba945c00f6d608d8975131ab3f25b22f2bcfe1dab221165194b2d4/itsdangerous-0.24.tar.gz"
+    url: ""
```
Comment (sschuberth) [inline analyzer/src/funTest/assets/projects/synthetic/python-inspector-expected-output.yml:105]: Why's this missing?

File: analyzer/src/funTest/assets/projects/synthetic/python-inspector-expected-output.yml [hunk 13]
```diff
@@ -160,27 +154,25 @@ packages:
...
     spdx_expression: "BSD-3-Clause"
     mapped:
       BSD: "BSD-3-Clause"
-      BSD License: "BSD-3-Clause"
-  description: "Implements a XML/HTML/XHTML Markup safe string for Python"
+  description: "MarkupSafe"
   homepage_url: "http://github.com/pallets/markupsafe"
   binary_artifact:
-    url: "https://files.pythonhosted.org/packages/4d/de/32d741db316d8fdb7680822dd37001ef7a448255de9699ab4bfcbdf4172b/MarkupSafe-1.0.tar.gz"
+    url: ""
```
Comment (sschuberth) [inline analyzer/src/funTest/assets/projects/synthetic/python-inspector-expected-output.yml:167]: Why's this missing?

File: analyzer/src/main/kotlin/managers/Pip.kt [hunk 14]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2017-2021 HERE Europe B.V.
+ * Copyright (C) 2017-2022 The ORT Project Authors (see <https://github.com/oss-review-toolkit/ort/blob/main/NOTICE>)
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/Pip.kt:2]: The new Copyright statement should just have the inception year.

File: analyzer/src/main/kotlin/managers/utils/PythonInspector.kt [hunk 15]
```diff
@@ -22,12 +22,28 @@ package org.ossreviewtoolkit.analyzer.managers.utils
...
+import org.ossreviewtoolkit.model.Package
+import org.ossreviewtoolkit.model.PackageReference
+import org.ossreviewtoolkit.model.RemoteArtifact
+import org.ossreviewtoolkit.model.VcsInfo
+import org.ossreviewtoolkit.model.VcsType
 import org.ossreviewtoolkit.utils.common.CommandLineTool
-import org.ossreviewtoolkit.utils.common.ProcessCapture
+import org.ossreviewtoolkit.utils.ort.createOrtTempFile
+
+private val json = Json { ignoreUnknownKeys = true }
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/utils/PythonInspector.kt:46]: This should be public, in a companion object, and be reused in the test (compare my earlier remark, and see existing similar code).

File: analyzer/src/main/kotlin/managers/utils/PythonInspector.kt [hunk 16]
```diff
@@ -36,18 +52,20 @@ internal object PythonInspector : CommandLineTool {
 
     override fun getVersionRequirement(): Requirement = Requirement.buildIvy("[0.7.1,)")
 
+    @OptIn(ExperimentalSerializationApi::class)
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/utils/PythonInspector.kt:55]: So now it probably pays off to do this on the `KotlinCompile` task level.

File: analyzer/src/main/kotlin/managers/utils/PythonInspector.kt [hunk 17]
```diff
@@ -113,3 +137,72 @@ internal class PythonInspectorParty(
...
+        val pkg = packages.first()
+
+        fun PythonInspectorPackage.getHash(): Hash =
+            when {
+                sha512 != null -> Hash(sha512, HashAlgorithm.SHA512)
+                sha256 != null -> Hash(sha256, HashAlgorithm.SHA256)
+                sha1 != null -> Hash(sha1, HashAlgorithm.SHA1)
+                md5 != null -> Hash(md5, HashAlgorithm.MD5)
+                else -> Hash.NONE
+            }
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/utils/PythonInspector.kt:161]: Maybe just `Hash.create(sha512 ?: sha256 ?: sha1 ?: md5 ?: "")`?

File: analyzer/src/main/kotlin/managers/utils/PythonInspector.kt [hunk 18]
```diff
@@ -113,3 +137,72 @@ internal class PythonInspectorParty(
...
+                )
+            } ?: RemoteArtifact.EMPTY
+
+        Package(
+            // The package has a namespace property which is currently always empty. Deliberately set the namespace to
+            // an empty string here to be consistent with the resolved packages which do not have a namespace property.
+            id = Identifier(type = TYPE, namespace = "", name = pkg.name, version = pkg.version),
+            purl = pkg.purl,
+            authors = pkg.parties.toAuthors(),
+            declaredLicenses = sortedSetOf(pkg.declaredLicense),
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/utils/PythonInspector.kt:177]: This apparently needs to be filtered for non-empty licenses. (Or fixed in python-inspector itself.)

File: analyzer/src/main/kotlin/PackageManager.kt [hunk 19]
```diff
@@ -318,6 +319,11 @@ abstract class PackageManager(
         }
     }
 
+    /**
+     * Get the [Options] from the [PackageManagerConfiguration] for this [package manager][managerName].
+     */
+    protected fun getOptions(): Options = analyzerConfig.getPackageManagerConfiguration(managerName)?.options.orEmpty()
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/PackageManager.kt:325]: How about making this take an optional key-argument, so one could write `getOptions(OPTION_DIRECT_DEPENDENCIES_ONLY)` instead of `getOptions()[OPTION_DIRECT_DEPENDENCIES_ONLY]`? Or make it a property?

File: analyzer/src/main/kotlin/managers/Pip.kt [hunk 20]
```diff
@@ -40,6 +40,7 @@ import org.ossreviewtoolkit.model.ProjectAnalyzerResult
 import org.ossreviewtoolkit.model.Scope
 import org.ossreviewtoolkit.model.VcsInfo
 import org.ossreviewtoolkit.model.config.AnalyzerConfiguration
+import org.ossreviewtoolkit.model.config.PackageManagerConfiguration
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/Pip.kt:40]: Commit message nit: s/linux/Linux/

File: analyzer/src/main/kotlin/managers/utils/PythonInspector.kt [hunk 21]
```diff
@@ -57,13 +57,17 @@ internal object PythonInspector : CommandLineTool {
         workingDir: File,
         definitionFile: File,
         pythonVersion: String = "38",
+        operatingSystem: String = "linux"
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/utils/PythonInspector.kt:60]: Shouldn't this use `OPTION_OPERATING_SYSTEM_DEFAULT`?

File: analyzer/src/funTest/kotlin/managers/PipFunTest.kt [hunk 22]
```diff
@@ -116,4 +117,14 @@ class PipFunTest : WordSpec({
     }
 })
 
-private fun createPip() = Pip("PIP", USER_DIR, DEFAULT_ANALYZER_CONFIGURATION, DEFAULT_REPOSITORY_CONFIGURATION)
+private fun createPip(pythonVersion: String = "3.10") =
```
Comment (sschuberth) [inline analyzer/src/funTest/kotlin/managers/PipFunTest.kt:120]: Shouldn't this use `OPTION_PYTHON_VERSION_DEFAULT`?

**Review by sschuberth — CHANGES_REQUESTED** [review]

**Review by fviernau — CHANGES_REQUESTED** [review]

File: analyzer/src/funTest/assets/projects/synthetic/pip-expected-output.yml (hunk 4 above)
Comment (mnonnenmacher) [reply analyzer/src/funTest/assets/projects/synthetic/pip-expected-output.yml:109]: I guess because PyPI only lists the .tar.gz URL for this package: https://pypi.org/pypi/itsdangerous/0.24/json
I will collect such issues and report them to python-inspector once I'm done with this PR and the upgrade to 0.8.2 which was released in the meantime.

File: analyzer/src/funTest/kotlin/managers/utils/PythonInspectorFunTest.kt [hunk 23]
```diff
@@ -0,0 +1,68 @@
...
+import kotlinx.coroutines.withContext
+import kotlinx.serialization.ExperimentalSerializationApi
+import kotlinx.serialization.json.Json
+import kotlinx.serialization.json.decodeFromStream
+
+import org.ossreviewtoolkit.utils.common.safeDeleteRecursively
+import org.ossreviewtoolkit.utils.test.createSpecTempDir
+
+@OptIn(ExperimentalSerializationApi::class)
+class PythonInspectorFunTest : StringSpec({
```
[the remark this answers is not in the document]
Comment (mnonnenmacher) [reply analyzer/src/funTest/kotlin/managers/utils/PythonInspectorFunTest.kt:30]: No, because not all keys of the python-inspector output are contained in the model.

File: analyzer/src/funTest/kotlin/managers/utils/PythonInspectorFunTest.kt [hunk 24]
```diff
@@ -0,0 +1,68 @@
...
+    val projectsDir = File("src/funTest/assets/projects").absoluteFile
+    val json = Json { ignoreUnknownKeys = true }
```
[the remark this answers is not in the document]
Comment (mnonnenmacher) [reply analyzer/src/funTest/kotlin/managers/utils/PythonInspectorFunTest.kt]: It's moved to `PythonInspector` in the next commit.

File: analyzer/src/funTest/kotlin/managers/utils/PythonInspectorFunTest.kt [hunk 25]
```diff
@@ -0,0 +1,68 @@
...
+
+    "python-inspector output can be deserialized" {
+        val definitionFile = projectsDir.resolve("synthetic/pip/requirements.txt")
+        val workingDir = definitionFile.parentFile
```
[the remark this answers is not in the document]
Comment (mnonnenmacher) [reply analyzer/src/funTest/kotlin/managers/utils/PythonInspectorFunTest.kt:35]: I think the test is still short, I would do this if another test is added to the file.

File: analyzer/src/funTest/kotlin/managers/utils/PythonInspectorFunTest.kt [hunk 26]
```diff
@@ -0,0 +1,68 @@
...
+                workingDir = workingDir,
+                outputFile = outputFile.absolutePath,
+                definitionFile = definitionFile,
+                pythonVersion = "27"
+            )
+        } finally {
+            workingDir.resolve(".cache").safeDeleteRecursively(force = true)
+        }
+
+        val result = withContext(Dispatchers.IO) {
```
[the remark this answers is not in the document]
Comment (mnonnenmacher) [reply analyzer/src/funTest/kotlin/managers/utils/PythonInspectorFunTest.kt]: I prefer to decode the input stream directly instead of reading the whole file into memory first.

File: analyzer/src/funTest/assets/projects/external/spdx-tools-python-expected-output.yml (hunk 1 above)
Comment (mnonnenmacher) [reply analyzer/src/funTest/assets/projects/external/spdx-tools-python-expected-output.yml:145]: See the response on another similar comment.

File: analyzer/src/funTest/assets/projects/synthetic/pip-expected-output.yml (hunk 5 above)
Comment (mnonnenmacher) [reply analyzer/src/funTest/assets/projects/synthetic/pip-expected-output.yml:165]: See the response on another similar comment.

File: analyzer/src/funTest/assets/projects/synthetic/pipenv-expected-output.yml (hunk 8 above)
Comment (mnonnenmacher) [reply analyzer/src/funTest/assets/projects/synthetic/pipenv-expected-output.yml:99]: See the response on another similar comment.

File: analyzer/src/funTest/assets/projects/synthetic/pipenv-expected-output.yml (hunk 9 above)
Comment (mnonnenmacher) [reply analyzer/src/funTest/assets/projects/synthetic/pipenv-expected-output.yml:163]: See the response on another similar comment.

File: analyzer/src/funTest/assets/projects/synthetic/poetry-expected-output.yml (hunk 10 above)
Comment (mnonnenmacher) [reply analyzer/src/funTest/assets/projects/synthetic/poetry-expected-output.yml:42]: See the response on another similar comment.

File: analyzer/src/funTest/assets/projects/synthetic/python-inspector-expected-output.yml (hunk 12 above)
Comment (mnonnenmacher) [reply analyzer/src/funTest/assets/projects/synthetic/python-inspector-expected-output.yml:99]: See the response on another similar comment.

File: analyzer/src/funTest/assets/projects/synthetic/python-inspector-expected-output.yml (hunk 13 above)
Comment (mnonnenmacher) [reply analyzer/src/funTest/assets/projects/synthetic/python-inspector-expected-output.yml:161]: See the response on another similar comment.

File: analyzer/src/main/kotlin/managers/utils/PythonInspector.kt (hunk 15 above)
Comment (mnonnenmacher) [reply analyzer/src/main/kotlin/managers/utils/PythonInspector.kt:44]: It's moved here from the test class in this commit, no code outside of this class will deserialize python-inspector output directly.

File: analyzer/src/main/kotlin/managers/utils/PythonInspector.kt [hunk 27]
```diff
@@ -113,3 +137,72 @@ internal class PythonInspectorParty(
     val email: String?,
     val url: String?
 )
+
+private const val TYPE = "PyPI"
+
+internal fun List<PythonInspectorPackage>.toOrtPackages(): SortedSet<Package> =
```
[the remark this answers is not in the document]
Comment (mnonnenmacher) [reply analyzer/src/main/kotlin/managers/utils/PythonInspector.kt:144]: No, they are used from the `Pip` class.

File: analyzer/src/funTest/assets/projects/external/example-python-flask-expected-output.yml [hunk 28]
```diff
@@ -33,26 +33,23 @@ packages:
 - id: "PyPI::click:8.1.3"
   purl: "pkg:pypi/click@8.1.3"
   authors:
-  - "Armin Ronacher"
+  - "Armin Ronacher <armin.ronacher@active-4.com>"
   declared_licenses:
-  - "BSD License"
```
[the remark this answers is not in the document]
Comment (mnonnenmacher) [reply analyzer/src/funTest/assets/projects/external/example-python-flask-expected-output.yml:38]: python-inspector is not using the classifiers to gather license information, see for example:

https://pypi.org/pypi/click/6.7/json
This does only have "BSD License" in the classifier but no license property.

https://pypi.org/pypi/rdflib/5.0.0/json
This does have a declared license "BSD-3-Clause" which I think we ignored in the old implementation, because we only used information from PyPI to fill empty fields.

I will create a feature request with python-inspector to provide the classifiers once I'm done with upgrading.

File: analyzer/src/main/kotlin/managers/utils/PythonInspector.kt [hunk 29]
```diff
@@ -60,3 +63,53 @@ internal object PythonInspector : CommandLineTool {
         return run(workingDir, *commandLineOptions.toTypedArray())
     }
 }
+
+@Serializable
+internal data class PythonInspectorResult(
```
[the remark this answers is not in the document]
Comment (mnonnenmacher) [reply analyzer/src/main/kotlin/managers/utils/PythonInspector.kt:93]: I like the idea, I will do this in a follow-up PR where I upgrade python-inspector again and more model changes are required, to not have to do the rebase work in this PR.

File: analyzer/src/main/kotlin/managers/Pip.kt (hunk 14 above)
Comment (mnonnenmacher) [reply analyzer/src/main/kotlin/managers/Pip.kt]: This is fixed after rebasing on your copyright cleanup PR.

File: analyzer/src/main/kotlin/PackageManager.kt (hunk 19 above)
Comment (mnonnenmacher) [reply analyzer/src/main/kotlin/PackageManager.kt]: I have made it a property.

File: analyzer/src/funTest/kotlin/managers/PipFunTest.kt (hunk 22 above)
Comment (mnonnenmacher) [reply analyzer/src/funTest/kotlin/managers/PipFunTest.kt:120]: I want the tests to be decoupled from the implementation in this regard, there will be a follow-up which cleans up the tests a bit, e.g. by running some test for linux and windows, and while at it I will consider if it makes sense to couple some tests to the default option value.

File: analyzer/src/funTest/assets/projects/external/example-python-flask-expected-output.yml [hunk 30]
```diff
@@ -66,27 +63,25 @@ packages:
...
-      BSD License: "BSD-3-Clause"
-  description: "A microframework based on Werkzeug, Jinja2 and good intentions"
+  description: "Flask"
   homepage_url: "http://github.com/pallets/flask/"
   binary_artifact:
     url: "https://files.pythonhosted.org/packages/0e/e9/37ee66dde483dceefe45bb5e92b387f990d4f097df40c400cf816dcebaa4/Flask-0.12-py2.py3-none-any.whl"
     hash:
-      value: "d3351b10f54446203ac0fd8839850c62"
-      algorithm: "MD5"
+      value: "7f03bb2c255452444f7265eddb51601806e5447b6f8a2d50bbc77a654a14c118"
```
[the remark this answers is not in the document]
Comment (mnonnenmacher) [reply analyzer/src/funTest/assets/projects/external/example-python-flask-expected-output.yml:78]: I have added a sentence to the commit message.

File: analyzer/src/funTest/assets/projects/external/example-python-flask-expected-output.yml [hunk 31]
```diff
@@ -33,26 +33,23 @@ packages:
 - id: "PyPI::click:8.1.3"
   purl: "pkg:pypi/click@8.1.3"
   authors:
-  - "Armin Ronacher"
+  - "Armin Ronacher <armin.ronacher@active-4.com>"
```
Comment (fviernau) [inline analyzer/src/funTest/assets/projects/external/example-python-flask-expected-output.yml:36]: commit-msg: "move **to** the strongest algorithm that is available."

Note: switching to a different hash algorithm could be factored out into a separate commit.

File: analyzer/src/main/kotlin/managers/Pip.kt [hunk 32]
```diff
@@ -115,10 +116,18 @@ object PythonVersion : CommandLineTool, Logging {
         }
 }
 
+private const val OPTION_OPERATING_SYSTEM = "operatingSystem"
+private const val OPTION_OPERATING_SYSTEM_DEFAULT = "linux"
+private val OPERATING_SYSTEMS = listOf("linux", "mac", "windows")
```
Comment (fviernau) [inline analyzer/src/main/kotlin/managers/Pip.kt:56]: Consider renaming `OPTION_OPERATING_SYSTEM_DEFAULT` to `OPTION_OPERATING_SYSTEM_LINUX`
to enable re-use in line 121 and maybe also next to @sschuberth 's comment below.

File: analyzer/src/funTest/assets/projects/external/example-python-flask-expected-output.yml (hunk 28 above)
[the remark this answers is not in the document]
Comment (fviernau) [reply analyzer/src/funTest/assets/projects/external/example-python-flask-expected-output.yml:38]: I've added the "breaking change" label, as classifiers are not used, so declared licenses may differ which in turn may break declared license mappings (ort configuration).

**Review by fviernau — APPROVED** [review]

File: analyzer/src/funTest/assets/projects/external/example-python-flask-expected-output.yml (hunk 28 above)
[the remark this answers is not in the document]
Comment (sschuberth) [reply analyzer/src/funTest/assets/projects/external/example-python-flask-expected-output.yml:38]: It would be nice to have the issue about parsing classifiers filed before the merge of this PR, so the issue could be referenced from the "Pip: Capture metadata using python-inspector" commit to explain the diff in the expected results.

File: analyzer/src/main/kotlin/managers/utils/PythonInspector.kt (hunk 21 above)
Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/utils/PythonInspector.kt:56]: How do we maintain this / recognize if the defaults of the python-inspector tool change?

Comment (mnonnenmacher) [reply analyzer/src/main/kotlin/managers/utils/PythonInspector.kt:56]: Manually, I guess. But maybe I remove those defaults altogether in a later iteration when refactoring the tests as mentioned in some other comment, because we might not need default values here anymore. In any way `PythonInspector` is a tool used by `Pip` so I wouldn't like it to directly depend on configuration options of `Pip`.

File: analyzer/src/funTest/assets/projects/external/example-python-flask-expected-output.yml (hunk 31 above)
Comment (mnonnenmacher) [reply analyzer/src/funTest/assets/projects/external/example-python-flask-expected-output.yml:36]: I don't want to spend the time on updating all the expected result files again twice.

File: analyzer/src/main/kotlin/managers/Pip.kt (hunk 32 above)
Comment (mnonnenmacher) [reply analyzer/src/main/kotlin/managers/Pip.kt:56]: I'd rather emphasize that this is the default value of the option than what the value is.

File: analyzer/src/funTest/assets/projects/external/example-python-flask-expected-output.yml (hunk 28 above)
[the remark this answers is not in the document]
Comment (mnonnenmacher) [reply analyzer/src/funTest/assets/projects/external/example-python-flask-expected-output.yml:38]: Done.

File: analyzer/src/funTest/assets/projects/external/example-python-flask-expected-output.yml (hunk 31 above)
Comment (mnonnenmacher) [reply analyzer/src/funTest/assets/projects/external/example-python-flask-expected-output.yml:36]: Commit message fixed.

File: analyzer/src/funTest/assets/projects/synthetic/pip-expected-output.yml (hunk 4 above)
Comment (sschuberth) [reply analyzer/src/funTest/assets/projects/synthetic/pip-expected-output.yml:103]: > we used the `.tar.gz` as the source artifact

You mean "binary artifact" here, I guess?

**Review by sschuberth — APPROVED** [review]

**Review by fviernau — APPROVED** [review]
